### PR TITLE
Lwm2m json lib integration

### LINKED ISSUE #43103: LwM2M library should use JSON library for parsing
**Is your enhancement proposal related to a problem? Please describe.**
This has been originally discussed in https://github.com/zephyrproject-rtos/zephyr/pull/41718.

Current implementation of LwM2M library, more specifically its JSON format parser and generator, directly consumes JSON format, and directly produces JSON format by using simple character based parsing. In the mentioned PR a new content format SenML-JSON is submitted, which follows the same model.

However, in Zephyr, there is already a [library for JSON parsing](https://docs.zephyrproject.org/latest/reference/misc/index.html#json) which should be used. This should make the code a bit easier to maintain, and in case multiple JSON formats are compiled in, a bit smaller in flash size as well.

**Describe the solution you'd like**
Use existing JSON library in LwM2M library. Make those JSON specific content formats then depend on Kconfig option [CONFIG_JSON_LIBRARY](https://docs.zephyrproject.org/latest/reference/kconfig/CONFIG_JSON_LIBRARY.html#std-kconfig-CONFIG_JSON_LIBRARY)

**Describe alternatives you've considered**
Alternative would be to use existing model of hard coding the JSON content format there. It is error prone and may leave some known bugs unfixed, if those are only fixed in library.

### PR DESCRIPTION
**Pull request by jheiskan81**
Update Zephyr Json lib API for integrate library for LwM2M JSON 1.0 and SenML-JSON module.

**New Types**
- Added support for Generic Numeric type (float, 64-bit)
- Added support for Opaque string type.
- Added support parse Array data in object for separate array parsing

**New API for Json Array parsing Object 1 by 1**
- json_arr_separate_object_parse_init() init array parse
- json_arr_separate_parse_object() Parsing 1 Json Object

SenML-JSON coap block-wise support is missed by using JSON library. Library is not support fragmented data.
LwM2M JSON 1.0 is not anymore default enabled.

Fixes #43103

### CODE REVIEW

**Comment by jheiskan81** [comment]
@SeppoTakalo @VeijoPesonen @jarlamsa @rlubos Please review.

**Review by SeppoTakalo — CHANGES_REQUESTED** [review]

File: subsys/net/lib/lwm2m/lwm2m_rw_senml_json.c [hunk 1]
```diff
@@ -17,262 +17,249 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME);
...
+	} obj;
+};
 
-	if (buf_read(attrbute_name, fd->name_len, CPKT_BUF_READ(in->in_cpkt), &fd->name_offset) <
-	    0) {
-		LOG_ERR("Error parsing attribute name!");
-		return SENML_JSON_UNKNOWN_ATTRIBUTE;
-	}
+struct json_out_formatter_data {
+	/* flags */
```
Comment (rlubos) [inline subsys/net/lib/lwm2m/lwm2m_rw_senml_json.c:89]: Do we need those comments? They don't add much value as they simply repeat the struct member name.

File: subsys/net/lib/lwm2m/lwm2m_rw_senml_json.c [hunk 2]
```diff
@@ -17,262 +17,249 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME);
...
-			}
+static const struct json_obj_descr senml_string_descr[] = {
+	JSON_OBJ_DESCR_PRIM_NAMED(struct senml_string_payload, "n",
+				  name, JSON_TOK_STRING),
+	JSON_OBJ_DESCR_PRIM_NAMED(struct senml_string_payload, "vs",
+				  val_string, JSON_TOK_STRING),
+};
 
-			break;
+struct json_in_formatter_data {
```
Comment (rlubos) [inline subsys/net/lib/lwm2m/lwm2m_rw_senml_json.c:227]: Unneeded empty line.

File: subsys/net/lib/lwm2m/lwm2m_rw_json.c [hunk 3]
```diff
@@ -68,188 +67,365 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME);
...
+				  val_bool, JSON_TOK_TRUE),
+	JSON_OBJ_DESCR_PRIM_NAMED(struct json_obj_struct, "ov",
+				  val_object_link, JSON_TOK_STRING),
+	JSON_OBJ_DESCR_PRIM_NAMED(struct json_obj_struct, "sv",
+				  val_string, JSON_TOK_STRING),
+};
 
-#define TOKEN_BUF_LEN	64
+#define	JSON_N_TYPE		1
+#define	JSON_V_TYPE		2
```
Comment (rlubos) [inline subsys/net/lib/lwm2m/lwm2m_rw_json.c:141]: Alignment

File: subsys/net/lib/lwm2m/lwm2m_rw_json.c [hunk 4]
```diff
@@ -68,188 +67,365 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME);
...
+				  name, JSON_TOK_STRING),
+	JSON_OBJ_DESCR_PRIM_NAMED(struct json_string_payload, "sv",
+				  val_string, JSON_TOK_STRING),
+};
 
 struct json_out_formatter_data {
-	/* offset position storage */
-	uint16_t mark_pos_ri;
 
 	/* flags */
```
Comment (rlubos) [inline subsys/net/lib/lwm2m/lwm2m_rw_json.c:179]: TBH, I'd just remove those undescriptive comments.

File: subsys/net/lib/lwm2m/lwm2m_rw_json.c [hunk 5]
```diff
@@ -68,188 +67,365 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME);
...
+
+	if (ret < 0) {
+		return ret;
+	}
+
+	return 0;
+}
+
+static int number_to_string(const char *format,
+			  ...)
```
Comment (rlubos) [inline subsys/net/lib/lwm2m/lwm2m_rw_json.c:219]: No need to split arguments.

**Review by rlubos — COMMENTED** [review]
Some nitpicking from my side.

File: include/zephyr/data/json.h [hunk 6]
```diff
@@ -596,6 +623,38 @@ int json_obj_parse(char *json, size_t len,
 int json_arr_parse(char *json, size_t len,
 	const struct json_obj_descr *descr, void *val);
 
+/**
+ * @brief Initialized JSON-encoded array data to @a json class for single object parsing by
+ * @a payload, with size @a len
+ *
+ * Function validate that Json Array start is detetcted and initialice @a json object for
```
Comment (SeppoTakalo) [inline include/zephyr/data/json.h:630]:
```suggestion
 * Function validate that Json Array start is detected and initialize @a json object for
```

File: include/zephyr/data/json.h [hunk 7]
```diff
@@ -596,6 +623,38 @@ int json_obj_parse(char *json, size_t len,
...
+int json_arr_separate_object_parse_init(struct json_obj *json, char *payload, size_t len);
+
+/**
+ * @brief Parses the JSON-encoded object pointed to by @a json object array, with
+ * size @a len, according to the descriptor pointed to by @a descr.
+ * Values are stored in a struct pointed to by @a val
+ *
+ * @param json Pointer to JSON-object message state
+ * @param descr Pointer to the descriptor array
+ * @param descr_len Number of elements in the descriptor array. Must be less
```
[the remark this answers is not in the document]
Comment (jheiskan81) [reply include/zephyr/data/json.h:649]: 31. It is depend library implementation

File: subsys/net/lib/lwm2m/lwm2m_rw_senml_json.c (hunk 1 above)
Comment (jheiskan81) [reply subsys/net/lib/lwm2m/lwm2m_rw_senml_json.c:89]: Ok I remove those.

File: subsys/net/lib/lwm2m/lwm2m_rw_senml_json.c [hunk 8]
```diff
@@ -17,262 +17,249 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME);
...
-			} else if (attrbute_name[1] == 's') {
-				return SENML_JSON_STRING_VALUE_ATTRIBUTE;
-			}
-		}
-	} else if (fd->name_len == 3) {
-		if (attrbute_name[0] == 'v' && attrbute_name[1] == 'l' && attrbute_name[2] == 'o') {
-			return SENML_JSON_OBJ_LINK_VALUE_ATTRIBUTE;
-		}
-	}
+	char bn_string[sizeof("/65535/999/") + 1];
```
[the remark this answers is not in the document]
Comment (jheiskan81) [reply subsys/net/lib/lwm2m/lwm2m_rw_senml_json.c:97]: Ah. Ok I will fix that.

File: subsys/net/lib/lwm2m/lwm2m_rw_senml_json.c [hunk 9]
```diff
@@ -17,262 +17,249 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME);
...
+};
 
-	if (block_ctx && block_ctx->json_flags) {
-		/* Store from last sequence */
-		fd->json_flags = block_ctx->json_flags;
-		block_ctx->json_flags = 0;
-	}
+#define	JSON_BN_TYPE	1
+#define	JSON_N_TYPE		2
+#define	JSON_V_TYPE		4
```
[the remark this answers is not in the document]
Comment (jheiskan81) [reply subsys/net/lib/lwm2m/lwm2m_rw_senml_json.c:135]: Ok I will fix.

**Comment by jheiskan81** [comment]
@SeppoTakalo and @rlubos I have fixed your comments

**Review by SeppoTakalo — APPROVED** [review]

**Review by rlubos — APPROVED** [review]

**Pull request merged**